### PR TITLE
[MNT] Fix warnings in widget tests

### PR DESCRIPTION
**Pull request by janezd**
##### Issue

Logs are polluted with warnings. Some are relevant and are fixed, and some are expected by tests and should be silenced. Logs are also printed out instead of being tested.

##### Includes
- [X] Code changes

### CODE REVIEW

File: Orange/widgets/data/owrank.py [hunk 1]
```diff
@@ -368,31 +371,33 @@ def set_learner(self, scorer, id):
...
+                try:
+                    scores = np.array([estimator(data, attr)
+                                       for attr in data.domain.attributes])
+                except ValueError:
+                    log.error("%s doesn't work on this data", method.name)
+                    scores = np.full(len(data.domain.attributes), np.nan)
+                else:
+                    log.warning("%s had to be computed separately for each "
+                                "variable", method.name)
+            return scores
```
Comment (janezd) [inline Orange/widgets/data/owrank.py]: Reorganized to give an error or a warning (or neither), not both.

File: Orange/widgets/data/owrank.py [hunk 2]
```diff
@@ -142,14 +142,17 @@ def headerData(self, section, orientation, role=Qt.DisplayRole):
...
-        else:
-            vmax = np.nanmax(values)
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", ".*All-NaN slice encountered.*",
+                                    RuntimeWarning)
+            try:
+                vmin = np.nanmin(values)
+                if np.isnan(vmin):
+                    raise TypeError
+            except TypeError:
```
Comment (lanzagar) [inline Orange/widgets/data/owrank.py]: I prefer explicit code that checks before doing something "stupid" instead of doing it first and catching exceptions.
If we had that here, we would also not need to suppress warnings...

I am guessing it would be enough to check `if np.isfinite(values).any()` and use the computed values, otherwise set to inf?
Shorter, clearer, explicit.
(and not tested, so maybe not enough :))

Sidenote: If sometime in the future, we pass a list of values (instead of np.array), the warning would change from
All-NaN slice encountered
to
All-NaN axis encountered
Just to keep in mind when setting this filter in functions that are otherwise fine with any iterable.

File: Orange/widgets/data/owrank.py [hunk 3]
```diff
@@ -368,31 +371,33 @@ def set_learner(self, scorer, id):
...
-        try:
-            scores = np.asarray(estimator(data))
-        except ValueError:
-            log.warning("Scorer %s wasn't able to compute all scores at once",
-                        method.name)
+        # These errors often happen, but they result in nans, which
+        # are handled correctly by the widget
+        with warnings.catch_warnings():
+            warnings.filterwarnings('ignore', '.*true_divide.*')
+            warnings.filterwarnings('ignore', '.*double_scalars.*')
```
Comment (lanzagar) [inline Orange/widgets/data/owrank.py]: I think it would be better to solve the problem of warnings at the source, i.e. in the scoring method that produces them, then in potentially several places where they are used.
Also, it is not hard to imagine that a general warning filter for all methods here can cover up an actual problem/bug. If not now, then maybe in the future.

File: Orange/widgets/data/tests/test_owdatasampler.py [hunk 4]
```diff
@@ -64,7 +66,9 @@ def select_sampling_type(self, sampling_type):
     def test_no_intersection_in_outputs(self):
         """ Check whether outputs intersect and whether length of outputs sums
         to length of original data """
-        self.send_signal("Data", self.zoo)
+        iris = Table("iris")
+        iris.X = np.arange(iris.X.size).reshape(iris.X.shape)
```
[the remark this answers is not in the document]
Comment (janezd) [reply Orange/widgets/data/tests/test_owdatasampler.py]: No, it's to avoid warning about less than 4 instances in some classes (see the comment below).

File: Orange/widgets/data/owrank.py (hunk 2 above)
Comment (janezd) [reply Orange/widgets/data/owrank.py]: I agree but we don't understand the code well enough to dare to remove catching the `TypeError`. Hence I only changed the code to filter only warnings only in `vmin = np.nanmin(values)`.

File: Orange/projection/manifold.py [hunk 5]
```diff
@@ -109,7 +109,7 @@ class MDS(SklProjector):
     __wraps__ = skl_manifold.MDS
     name = 'MDS'
 
-    def __init__(self, n_components=2, metric=True, n_init=4, max_iter=300,
+    def __init__(self, n_components=2, metric=True, n_init=None, max_iter=300,
```
Comment (lanzagar) [inline Orange/projection/manifold.py]: I would prefer if our wrappers retain the same signature/defaults as the wrapped function as much as possible.
It is then much easier to update them, if/when these values change upstream (and they often do).

This leaves us with a couple of choices of how to combat the current awkward defaults:
- Fix the calls of MDS, and explicitly set `n_init` to 1 when `init` is used. Not sure it that is how the authors imagined it... (does not sound appealing)
- Let sklearn change it and filter out the warning when calling fit, if that bothers us
- We make the same change before them in our call. Basically what you did in L150, but do it when `init` is given instead of when `n_init` is not.

Comment (janezd) [reply Orange/projection/manifold.py]: The first option was the simplest because this happened only in one place.

File: Orange/widgets/unsupervised/owtsne.py [hunk 6]
```diff
@@ -169,12 +171,16 @@ def error(err):
                 error(self.Error.not_enough_rows)
             elif not self.data.domain.attributes:
                 error(self.Error.no_attributes)
-            elif not self.data.is_sparse() and \
-                    not np.sum(np.nanstd(self.data.X, axis=0)):
-                error(self.Error.constant_data)
             elif not self.data.is_sparse() and \
                     np.all(~np.isfinite(self.data.X)):
                 error(self.Error.no_valid_data)
+            elif not self.data.is_sparse():
```
Comment (lanzagar) [inline Orange/widgets/unsupervised/owtsne.py]: If we now have an `elif not sparse` then the above elif can be moved inside this one...

File: Orange/widgets/visualize/owboxplot.py [hunk 7]
```diff
@@ -680,16 +680,19 @@ def compute_tests(self):
...
+
+        # pylint: disable=comparison-with-itself
         def stat_ttest():
             d1, d2 = self.stats
-            if d1.n == 0 or d2.n == 0:
+            if d1.n < 2 or d2.n < 2:
                 return np.nan, np.nan
             pooled_var = d1.var / d1.n + d2.var / d2.n
+            # pylint: disable=comparison-with-itself
+            if pooled_var == 0 or pooled_var != pooled_var:
```
Comment (lanzagar) [inline Orange/widgets/visualize/owboxplot.py]: Isn't `np.isnan` clearer?

File: Orange/widgets/visualize/owboxplot.py [hunk 8]
```diff
@@ -724,15 +729,21 @@ def stat_ANOVA():
                 # t = "Mann-Whitney's z: %.1f (p=%.3f)" % (z, self.p)
             else:
                 t, self.p = stat_ttest()
-                t = "Student's t: %.3f (p=%.3f)" % (t, self.p)
+                if t == t:
```
Comment (lanzagar) [inline Orange/widgets/visualize/owboxplot.py]: Isn't `np.isnan` clearer?

File: Orange/widgets/visualize/owboxplot.py [hunk 9]
```diff
@@ -724,15 +729,21 @@ def stat_ANOVA():
...
+                    t = 0
         else:
             if self.compare == OWBoxPlot.CompareMedians:
                 t = ""
                 # U, self.p = -1, -1
                 # t = "Kruskal Wallis's U: %.1f (p=%.3f)" % (U, self.p)
             else:
                 F, self.p = stat_ANOVA()
-                t = "ANOVA: %.3f (p=%.3f)" % (F, self.p)
+                if F == F:
```
Comment (lanzagar) [inline Orange/widgets/visualize/owboxplot.py]: Isn't `np.isnan` clearer?

File: Orange/widgets/visualize/owboxplot.py [hunk 10]
```diff
@@ -724,15 +729,21 @@ def stat_ANOVA():
                 # t = "Mann-Whitney's z: %.1f (p=%.3f)" % (z, self.p)
             else:
                 t, self.p = stat_ttest()
-                t = "Student's t: %.3f (p=%.3f)" % (t, self.p)
+                if t == t:
+                    t = "Student's t: %.3f (p=%.3f)" % (t, self.p)
+                else:
+                    t = 0
```
Comment (lanzagar) [inline Orange/widgets/visualize/owboxplot.py]: `t = ""`
As in ANOVA below and CompareMedians...

File: Orange/widgets/visualize/owsieve.py [hunk 11]
```diff
@@ -43,8 +43,9 @@ def __init__(self, data, attr1, attr2):
         self.probs_x = self.observed.sum(axis=0) / self.n
         self.probs_y = self.observed.sum(axis=1) / self.n
         self.expected = np.outer(self.probs_y, self.probs_x) * self.n
-        self.residuals = \
-            (self.observed - self.expected) / np.sqrt(self.expected)
+        with np.errstate(divide="ignore", invalid="ignore"):
+            self.residuals = \
+                (self.observed - self.expected) / np.sqrt(self.expected)
```
Comment (lanzagar) [inline Orange/widgets/visualize/owsieve.py]: Can we not divide by zero instead of ignoring warnings?

File: Orange/classification/naive_bayes.py [hunk 12]
```diff
@@ -48,22 +49,62 @@ def __init__(self, log_cont_prob, class_prob, domain):
 
     def predict_storage(self, data):
         if isinstance(data, Instance):
-            data = [data]
-        if len(data.domain.attributes) == 0:
-            probs = np.tile(self.class_prob, (len(data), 1))
+            data = np.atleast_2d(data.X)
+        elif isinstance(data, Storage):
+            data = data.X
```
Comment (lanzagar) [inline Orange/classification/naive_bayes.py]: The `Storage` class does not have `X` - `Table` has it, and even there it is not declared properly :)
Joking aside, our Learners and Models can implement fit/predict which work with numpy arrays or fit_storage/predict_storage which do not assume/require numpy data.
Instead of `data = data.X` you could replace the implementation of predict_storage with the `predict` method. That can have implications for non-numpy based storages (SqlTable)...
In short - this change needs some consideration and should probably be in its own PR.

File: Orange/widgets/visualize/owscatterplot.py [hunk 13]
```diff
@@ -240,17 +241,22 @@ def effective_variables(self):
...
-        self.vizrank_button.setToolTip(text)
+        with warnings.catch_warnings():
+            # nanstd may trigger this warning, but it's harmless
+            warnings.filterwarnings("ignore", "Degrees of freedom .*",
+                                    RuntimeWarning)
+            self.vizrank.initialize()
+            is_enabled = self.data is not None and \
+                not self.data.is_sparse() and \
+                len(self.xy_model) > 2 and len(self.data[self.valid_data]) > 1 \
+                and np.all(np.nan_to_num(np.nanstd(self.data.X, 0)) != 0)
```
[the remark this answers is not in the document]
Comment (janezd) [reply Orange/widgets/visualize/owscatterplot.py]: You're right, I removed this from the commit.

File: Orange/widgets/visualize/owscatterplotgraph.py [hunk 14]
```diff
@@ -864,7 +864,10 @@ def _get_continuous_colors(self, c_data, subset):
         except the former is darker. If the data has a subset, the brush
         is transparent for points that are not in the subset.
         """
-        self.scale = DiscretizedScale(np.nanmin(c_data), np.nanmax(c_data))
+        with warnings.catch_warnings():
+            # DiscretizedScale handle's nans just fine
+            warnings.filterwarnings("ignore", "All-NaN slice.*", RuntimeWarning)
+            self.scale = DiscretizedScale(np.nanmin(c_data), np.nanmax(c_data))
```
[the remark this answers is not in the document]
Comment (janezd) [reply Orange/widgets/visualize/owscatterplotgraph.py]: Done.

File: Orange/widgets/visualize/owsieve.py (hunk 11 above)
Comment (janezd) [reply Orange/widgets/visualize/owsieve.py]: No, it's a vector. I could create a mask, but it would be annoying. It's simpler this way.

File: Orange/classification/naive_bayes.py (hunk 12 above)
Comment (janezd) [reply Orange/classification/naive_bayes.py]: Moved to #3540.

File: Orange/statistics/util.py [hunk 15]
```diff
@@ -353,7 +354,10 @@ def weighted_mean():
 
 def _nan_min_max(x, func, axis=0):
     if not sp.issparse(x):
-        return func(x, axis=axis)
+        with warnings.catch_warnings():
```
Comment (lanzagar) [inline Orange/statistics/util.py]: This currently still changes the behaviour of nanmin/nanmax compared to numpy's version.
I thought we agreed that for numpy arrays our functions should be equivalent to np.nanmin/max
